source/scheduler.hpp

```diff
@@ -24,7 +24,16 @@ class Scheduler {
         return instance;
     }
 
-    void add(std::unique_ptr<Event> event);
+    template <typename TEvent, typename... Args>
+    void add(Args&&... args) {
+        static_assert(std::is_constructible_v<TEvent, Args&&...>,
+                      "Event must be constructible from given args");
+        static_assert(std::is_base_of_v<Event, TEvent>,
+                      "TEvent must be inherit from Event");
+
+        m_events.push(std::make_unique<TEvent>(args...));
```

Comment:
Is `emplace` a better choice here?
```suggestion
        m_events.emplace(std::make_unique<TEvent>(args...));
```